### PR TITLE
Transaction support

### LINKED ISSUE #27: Transactions support
Asto should support transactions in order to guarantee atomic updates of several keys.

### PR DESCRIPTION
**Pull request by Sammers21**
fixes #27

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#28](https://github.com/artipie/asto/issues/28) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/7eda327b-c623-4480-b4d8-b35b16d89e95); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by amihaiemil** [comment]
@Sammers21 I'll look over this in a few minutes. Until then:

1. Please don't use keywords such as ``fixes`` or ``closes`` in the description. One merged PR does not necessarily mean that the Issues is fixed and should be closed.

2. Please add a meaningful and structured description to your PR: what exactly are the changes/fixes?

File: src/main/java/com/artipie/asto/Storage.java [hunk 1]
```diff
@@ -74,4 +75,12 @@
...
      */
     CompletableFuture<Flow.Publisher<Byte>> value(Key key);
+
+    /**
+     * Start a transaction with specified keys.
+     *
+     * @param keys The keys regarding which transaction is atomic
+     * @return transaction
+     */
+    CompletableFuture<TransactionStorage> transaction(List<Key> keys);
```
Comment (amihaiemil) [inline src/main/java/com/artipie/asto/Storage.java]: @Sammers21 Please specify a more detailed JavaDoc. For example, does this method start a transaction for each ``Key``?

File: src/main/java/com/artipie/asto/TransactionStorage.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Transaction.
+ *
+ * @since 0.10.0
+ */
+public interface TransactionStorage extends Storage {
```
Comment (amihaiemil) [inline src/main/java/com/artipie/asto/TransactionStorage.java]: @Sammers21 Please add a more detailed JavaDoc. What exactly is a ``Transaction``? 

Also, maybe we can remove the ``Storage`` suffix? Since this interface extends interface ``Storage`` it seems already obvious to me, that it is a Storage Transaction :)

File: src/main/java/com/artipie/asto/TransactionStorage.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    /**
+     * Keys the transaction performed on.
+     * @return The keys
```
Comment (amihaiemil) [inline src/main/java/com/artipie/asto/TransactionStorage.java]: @Sammers21 Also a more detailed JavaDoc here. What are the ``keys``? Where do they come from?

**Comment by amihaiemil** [comment]
@Sammers21 Please also fix the Travis CI build. Everything should be Green before a PR is merged.

**Comment by Sammers21** [comment]
@amihaiemil , sorry, this PR is not ready for review. I should be creating a draft one.

**Comment by amihaiemil** [comment]
@Sammers21 continue working on this, no problem. 

I don't think we have the concept of Draft PR in Zerocracy, so I start reviewing as soon as I get assigned :)

**Comment by Sammers21** [comment]
@amihaiemil, Draft PR's is a GitHub concept: https://github.blog/2019-02-14-introducing-draft-pull-requests/

**Comment by Sammers21** [comment]
@amihaiemil , you can review the Pull Request.

**Comment by Sammers21** [comment]
> Please don't use keywords such as fixes or closes in the description. One merged PR does not necessarily mean that the Issues is fixed and should be closed.

In this case, it means that the issue is fixed by this Pull Request.

File: src/main/java/com/artipie/asto/TransactionStorage.java (hunk 2 above)
Comment (Sammers21) [reply src/main/java/com/artipie/asto/TransactionStorage.java]: > Also, maybe we can remove the Storage suffix? Since this interface extends interface Storage it seems already obvious to me, that it is a Storage Transaction :)

Agree

File: src/main/java/com/artipie/asto/TransactionStorage.java (hunk 3 above)
Comment (Sammers21) [reply src/main/java/com/artipie/asto/TransactionStorage.java]: > Where do they come from?

Does it matter?

**Comment by Sammers21** [comment]
@amihaiemil , can you please review the Pull Request again?

**Comment by amihaiemil** [comment]
@Sammers21 sure, give me a few minutes :)

File: src/main/java/com/artipie/asto/rx/RxStorage.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import io.reactivex.rxjava3.core.Single;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * A reactive version of {@link com.artipie.asto.Storage}.
+ *
+ * @since 0.10
+ */
+public interface RxStorage {
```
Comment (amihaiemil) [inline src/main/java/com/artipie/asto/rx/RxStorage.java]: @Sammers21 If this a more specialized kind of Storage, shouldn't this interface extend interface Storage?

**Comment by Sammers21** [comment]
> Please make PRs with maximum 5-6 changed files, including tests. PR with 13 files is huge.

The diff is huge, the amount of effective changes is close to ~100 lines of code. 

> Please cover only one topic per PR. In this PR you covered at least 2 topics:
> 
> you introduced Storage and transaction
> you added Reactive implementation fors them.

As you might notice, the reactive wrapper has been in the repo before. It is just not reasonable to make the PR contain fewer changes, it is already as small as possible

**Comment by amihaiemil** [comment]
@Sammers21 ok, fine by me. I approved the PR, I'm not sure why there is a red X still... let me try again.

**Comment by amihaiemil** [comment]
@Sammers21 apparently it's because I don't have Write Access. 
@g4s8 should have a look and merge it, I believe. 

You are still missing the unit tests, though. Next time, please don't forget about them.

**Comment by Sammers21** [comment]
> There are no unit tests. How do you know your classes/objects work as expected?

We already have a test for FileStorage. Testing one-line-of-code wrappers are overkill.

**Comment by amihaiemil** [comment]
@Sammers21 
> Testing one-line-of-code wrappers are overkill.

I understand. Most developers from the industry would agree.

But last time I worked with this team, they had really high standards (code coverage > 80%). I'm not sure if this has changed or not, I wasn't around in the last period of time :)

**Comment by amihaiemil** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/artipie/asto/issues/28#issuecomment-580164859))

@amihaiemil This is what I know about this job in [CT2E6TK9B](https://www.0crat.com/p/CT2E6TK9B), as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#28](https://github.com/artipie/asto/issues/28) is [in scope](http://www.zerocracy.com/policy.html#14) for 2days
  * The role is `REV`
  * The job is not assigned to anyone
  * The [budget](http://www.zerocracy.com/policy.html#4) is not set yet
  * These users are banned and won't be assigned:
    * @sammers21[/z](https://www.0crat.com/u/sammers21): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/CT2E6TK9B?q=%7Bjob%3A%27gh%3Aartipie%2Fasto%2328%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by 0crat** [comment]
Job audit: `ARC` didn't merge or close PR

**Comment by 0crat** [comment]
Order was finished: +20 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Job was finished in 45 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)
